drivers/wifi/infineon/airoc_whd_hal_spi.c

```diff
@@ -0,0 +1,289 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <airoc_wifi.h>
+#include "airoc_whd_hal_common.h"
+
+#include <bus_protocols/whd_bus_spi_protocol.h>
+#include <bus_protocols/whd_spi.h>
+#include <whd_wifi_api.h>
+#include <zephyr/drivers/spi.h>
+
+LOG_MODULE_DECLARE(infineon_airoc_wifi, CONFIG_WIFI_LOG_LEVEL);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct whd_bus_priv {
+	whd_spi_config_t spi_config;
+	whd_spi_t spi_obj;
+};
+
+static whd_init_config_t init_config_default = {.thread_stack_size = CY_WIFI_THREAD_STACK_SIZE,
+						.thread_stack_start = NULL,
+						.thread_priority =
+							(uint32_t)CY_WIFI_THREAD_PRIORITY,
+						.country = CY_WIFI_COUNTRY};
+
+/******************************************************
+ *		Function
+ ******************************************************/
+
+int airoc_wifi_init_primary(const struct device *dev, whd_interface_t *interface,
+				whd_netif_funcs_t *netif_funcs,
+				whd_buffer_funcs_t *buffer_if)
+{
+	struct airoc_wifi_data *data = dev->data;
+	const struct airoc_wifi_config *config = dev->config;
+
+	whd_spi_config_t whd_spi_config = {
+		.is_spi_normal_mode = WHD_FALSE,
+	};
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_host_wake_gpios)
+	whd_oob_config_t oob_config = {
+		.host_oob_pin = (void *)&config->wifi_host_wake_gpio,
+		.dev_gpio_sel = DEFAULT_OOB_PIN,
+		.is_falling_edge =
+			(CY_WIFI_HOST_WAKE_IRQ_EVENT == GPIO_INT_TRIG_LOW) ? WHD_TRUE : WHD_FALSE,
+		.intr_priority = CY_WIFI_OOB_INTR_PRIORITY};
+	whd_spi_config.oob_config = oob_config;
+#endif
+#if defined(SPI_DATA_IRQ_SHARED)
+	data->prev_irq_state = 0;
+#endif
+
+	/* Pull bus select line low before enabling WiFi chip */
+	gpio_pin_configure_dt(&config->bus_select_gpio, GPIO_OUTPUT_INACTIVE);
+	gpio_pin_set_dt(&config->bus_select_gpio, 0);
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on returns fail");
+		return -ENODEV;
+	}
+
+	if (!spi_is_ready_dt(&config->bus_dev.bus_spi)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	/* Init wifi host driver (whd) */
+	cy_rslt_t whd_ret = whd_init(&data->whd_drv, &init_config_default,
+					&resource_ops, buffer_if, netif_funcs);
+	if (whd_ret == CY_RSLT_SUCCESS) {
+		whd_ret = whd_bus_spi_attach(data->whd_drv, &whd_spi_config,
+					    (whd_spi_t)&config->bus_dev.bus_spi);
+
+		if (whd_ret == CY_RSLT_SUCCESS) {
+			whd_ret = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (whd_ret != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
+			return -ENODEV;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Implement SPI Transfer wrapper
+ */
+
+whd_result_t whd_bus_spi_transfer(whd_driver_t whd_driver, const uint8_t *tx, size_t tx_length,
+				  uint8_t *rx, size_t rx_length, uint8_t write_fill)
+{
+	const struct spi_dt_spec *spi_obj = whd_driver->bus_priv->spi_obj;
+	int ret;
+	whd_result_t whd_ret = WHD_SUCCESS;
+
+#if defined(SPI_DATA_IRQ_SHARED)
+	ret = whd_bus_spi_irq_enable(whd_driver, false);
```

Comment:
SPI operations needs to be done with interrupt disabled?

Reply:
Yes, if the same SPI line is being used for data transfers and interrupts, which is the case on the Raspberry Pi Pico W.  Interrupts need to be disabled so data transfers don't trigger the interrupt.

Reply:
Driver is specific to board / SoC then? It shouldn't AFAIK.

Reply:
No, though it currently applies to just two boards, the Raspberry Pi Pico W and Pico 2 W.  Given Zephyr's code layers, the SPI driver has no way to know the WiFi module has this wacky wiring configuration - nor should it.  It's specific to WiFi on these boards, so it's the WiFi driver layer that has to deal with the issue, but anything in the WiFi driver to do so should otherwise be board-agnostic.  This was the best thing I could think of to encapsulate the problem.

Reply:
If this modules is going to be used in other boards, then it makes sense to add an additional quirk from dts (something like shared-interrupt) and differentiate it. Otherwise this module is bound to RPi pico families.

Reply:
I believe you're suggesting something like the property `spi-data-irq-shared`, which is defined in dts/bindings/wifi/infineon,airoc-wifi-spi.yaml?

Reply:
Oh, I missed that. yes can `spi-data-irq-shared` help make this part of code independent of Pico?

Reply:
That was my goal:  any device that needs to use a single pin for data input and interrupts from the WiFi module can use that property.  You can see it used in the board definition for the Pico W:  boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w.dts.

Originally I had Pico-specific code here, and previous reviewers pointed out how inappropriate that was, leading me to figure out a platform-independent way to trigger the necessary logic.

Reply:
But still the underlying idea about disabling interrupt in SPI as it's shared with the data pin is here?

Reply:
I don't understand your question, so let me restate the problem and my solution:  If the same GPIO pin is used for both data and interrupts coming from the WiFi module, interrupts have to be disabled during data transfers to prevent the data on the SPI input from triggering interrupts.  That's not a function of the SPI bus layer, it belongs in the layer managing WiFi transfers;  here, that's the Infineon AIROC driver.  Though there are currently only two boards that have this issue, the Raspberry Pi Pico W and Pico 2 W, I tried to implement this in a way that wasn't specific to those boards.

Can you explain what your concern is about this solution?

Reply:
You have the conditional ennoblement of interrupt using `#if defined(SPI_DATA_IRQ_SHARED)`, but do we still need to disable the interrupt even if data and interrupt is not shared -> https://github.com/zephyrproject-rtos/zephyr/pull/78773/files#diff-a359abf10a98428aea5f07434fb65f88e4544a4ddf9d26029ee3739c65fedea6R271 ?

Reply:
I think I understand your question now.  Enabling and disabling the IRQ line is normally under control of the driver code in the Infineon HAL module (modules/hal/infineon).  The function whd_bus_spi_irq_enable() (and the corresponding whd_bus_sdio_irq_enable() and whd_bus_m2m_irq_enable()) in the drivers/wifi/infineon layer is the platform-specific bit to do so.  That logic is independent of whether or not the GPIO is shared.

I think the board designers at Raspberry Pi were getting a bit desperate to squeeze the CYW43439 onto their board without disrupting the rest of their I/O design.  I hope no other board designer has to deal with a situation like this.